crates/bevy_ecs/src/storage/sorted_vec_set.rs

```diff
@@ -0,0 +1,250 @@
+use core::cmp::Ordering;
+use smallvec::SmallVec;
+
+/// Stores a sorted list of indices with quick implementation for union, difference, intersection.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct SortedVecSet<const N: usize>(SmallVec<[usize; N]>);
+
+impl<const N: usize> IntoIterator for SortedVecSet<N> {
+    type Item = usize;
+    type IntoIter = <SmallVec<[usize; N]> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<const N: usize> Default for SortedVecSet<N> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<const N: usize> SortedVecSet<N> {
+    /// Construct an empty vector
+    pub(crate) const fn new() -> Self {
+        Self(SmallVec::new_const())
+    }
+
+    /// Construct a new `SortedSmallVec` from a `Vec<usize>`.
+    ///
+    /// Elements are copied and put in a sorted order if the original `Vec` isn't ordered.
+    /// Duplicates are removed.
+    #[allow(dead_code)]
+    pub(crate) fn from_vec(vec: Vec<usize>) -> Self {
+        let mut sorted_vec = Self(SmallVec::with_capacity(vec.len()));
+        for value in vec {
+            sorted_vec.insert(value);
+        }
+        sorted_vec
+    }
+
+    pub(crate) fn iter(&self) -> impl Iterator<Item = usize> + '_ {
+        self.0.iter().copied()
+    }
+
+    /// Insert the value if it's not already present in the vector.
+    /// Maintains a sorted order.
+    pub(crate) fn insert(&mut self, index: usize) {
+        match self.0.binary_search(&index) {
+            // element already present in the vector
+            Ok(_) => {}
+            Err(pos) => {
+                self.0.insert(pos, index);
+            }
+        }
+    }
+
+    /// Removes a value if it's present in the vector
+    pub(crate) fn remove(&mut self, index: usize) {
+        if let Ok(pos) = self.0.binary_search(&index) {
+            self.0.remove(pos);
+        }
+    }
+
+    /// Returns true if the vector contains the value.
+    pub(crate) fn contains(&self, index: usize) -> bool {
+        self.0.binary_search(&index).is_ok()
+    }
+
+    /// Returns true if the vector is empty.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Empties the contents of the vector
+    pub(crate) fn clear(&mut self) {
+        self.0.clear();
+    }
+
+    /// Returns the number of elements in the vector.
+    pub(crate) fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub(crate) fn difference_with(&mut self, other: &Self) {
+        let mut j = 0;
+        self.0.retain(|current| {
+            // Advance past any smaller elements in other
+            while j < other.len() && other.0[j] < *current {
+                j += 1;
+            }
+            // It's only in the difference if it's not in other,
+            // and this is the only place in other it could be
+            j < other.len() && other.0[j] != *current
```

Comment:
I think you need to invert the entire condition for `difference_with`.  For example, if `other` is empty then we want to keep everything and should always return `true`, but `j < other.len()` would be false.  

```suggestion
            j == other.len() || other.0[j] != *current
```